Allow custom non-damage weaknesses, fix Blessed Counterstrike weakness

xyzzy42 opened the pull request:
Currently non-damage weaknesses are limited to a fixed set hard-coded in the
system with NON_DAMAGE_WEAKNESSES.

Add an optional field so the non-damage weakness type can be set.  If the field
is omitted, it will default to true if the weakness type is in
NON_DAMAGE_WEAKNESSES, false otherwise.

Address issue #16190

This allows creating custom non-damage weaknesses.  E.g., a definition of
"origin:signature:{item|origin.signature}" would create a weakness to strikes
from a specific player who is the origin of the effect with the weakness.  By
setting nonDamage to true, the weakness will only trigger once for a damage
roll, rather than triggering on every damage instance in the roll.

It also makes it possible to create weakness to spells/items with a trait or
traits that do not trigger multiple times on spells with multiple damage types.
This already exists for the magic traditions and a few other traits, which are
in the NON_DAMAGE_WEAKNESSES set, but not arbitrary traits.  E.g., there's no
way to make a weakness to fear spells that does not trigger twice on damage from
Invoke Spirits, which deals both mental and void damage.

xyzzy42 committed:
Fix Blessed Counterstrike effect's weakness

This should not trigger for each damage instance.  Addresses
issue #18859.  ￼

stwlam commented:
I think we're going to reframe this whole thing as apply-once weaknesses

stwlam commented on packs/feat-effects/effect-blessed-counterstrike.json:
```diff
@@ -28,6 +28,7 @@
                 ],
                 "key": "Weakness",
                 "label": "PF2E.IWR.Custom.EnemyStrikes",
+                "nonDamage": true,
```
A bit strange to classify a strike as non-damaging.

xyzzy42 replied:
I think of as it a weakness to the contact from being struck rather than as a weakness to the damage the strike does.

stwlam commented on src/module/actor/data/iwr.ts:
```diff
@@ -301,9 +301,12 @@ class Weakness extends IWR<WeaknessType> implements WeaknessSource {
     override value: number;
 
-    constructor(data: IWRConstructorData<WeaknessType> & { value: number }) {
+    readonly nonDamage: boolean;
+
+    constructor(data: IWRConstructorData<WeaknessType> & { value: number; nonDamage?: boolean }) {
         super(data);
         this.value = data.value;
+        this.nonDamage = data.nonDamage ?? NON_DAMAGE_WEAKNESSES.has(this.type);
     }
```
This needs to be part of the model definition

stwlam commented on src/module/rules/rule-element/iwr/weakness.ts:
```diff
@@ -12,6 +13,7 @@ class WeaknessRuleElement extends IWRRuleElement<WeaknessRuleSchema> {
             ...super.defineSchema(),
             value: new ResolvableValueField({ required: true, nullable: false, initial: undefined }),
             exceptions: this.createExceptionsField(this.dictionary),
+            nonDamage: new fields.BooleanField({ required: false, initial: undefined }),
```
Besides reframing, this needs to be locked down such that only custom weaknesses and ones already considered single-application can be marked as single-application.

xyzzy42 replied:
Should the new name be `applyOnce` or `singleApplication`?  

I can do that, but why is it necessary?  If someone wants to make an apply once weakness that is the same as an existing named weakness, why force it to be custom?

An example might be the existing `axe-vulnerability` weakness, which is I think supposed to be a apply-once weakness.  If someone wants the rule to be that way, why not let them change it?

xyzzy42 replied:
Should `applyOnce` of any value be rejected for non-custom weaknesses?  Or should it instead be required that the property, if present, must match the pre-defined apply onceness of the weakness.  So holy + applyOnce:true is ok, and fire + applyOnce:false is ok.

It'd be simpler, if there must be a restriction, to just only allow it for custom.  Otherwise you can't validate fail immediately since the values of either `type` or `applyOnce` might not be resolved.

xyzzy42 committed:
Add applyOnce field to Weakness RE

Currently non-damage weaknesses are limited to a fixed set hard-coded in the
system with NON_DAMAGE_WEAKNESSES.

Add an optional field so weaknesses can be set to apply once.  If the field
is omitted, it will default to true if the weakness type was in
NON_DAMAGE_WEAKNESSES, false otherwise.

"non-damage" is renamed to "apply once".

Address issue #16190

This allows creating custom non-damage weaknesses.  E.g., a definition of
"origin:signature:{item|origin.signature}" would create a weakness to strikes
from a specific player who is the origin of the effect with the weakness.  By
setting applyOnce to true, the weakness will only trigger once for a damage
roll, rather than triggering on every damage instance in the roll.

It also makes it possible to create weakness to spells/items with a trait or
traits that do not trigger multiple times on spells with multiple damage types.
This already exists for the magic traditions and a few other traits, which are
in the NON_DAMAGE_WEAKNESSES set, but not arbitrary traits.  E.g., there's no
way to make a weakness to fear spells that does not trigger twice on damage from
Invoke Spirits, which deals both mental and void damage.

xyzzy42 commented:
I have renamed to `applyOnce`.

`applyOnce` can now only be specified for custom weaknesses.

Since the weakness type(s) can be an array, it doesn't make sense to also supply `applyOnce`, as it would not be able to match multiple types, some of which might be apply once and some not.  This is not a problem for custom weaknesses, as custom is already required to be a single weakness per RE.

stwlam committed:
Fix compilation errors
